File: Data Structures & Algorithms/combination-target-sum/submission-18.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) { 
        const res = [] 
        const cur = [] 

        this.dfs(0, nums, target, 0, cur, res) 

        return res
    } 

    dfs(i, nums, target, total, cur, res) { 
        if (total === target) {
            res.push([...cur]) 
            return
        }  

        if (i >= nums.length || total > target) {
            return
        }

        cur.push(nums[i]) 
        this.dfs(i, nums, target, total + nums[i], cur, res) 
        cur.pop() 
        this.dfs(i + 1, nums, target, total, cur, res)
    }
}

File: Data Structures & Algorithms/permutations/submission-18.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        numSet = set(nums)  
        res = []

        def dfs(i, cur): 
            if i == len(nums): 
                res.append(cur.copy()) 
                return 

            for num in list(numSet):
                cur.append(num)
                numSet.remove(num)

                dfs(i + 1, cur)

                cur.pop()
                numSet.add(num)

        dfs(0, []) 
        return res